File: common/src/main/java/generations/gg/generations/core/generationscore/network/packets/dialogue/S2CChooseDialoguePacket.java
package generations.gg.generations.core.generationscore.network.packets.dialogue;

import generations.gg.generations.core.generationscore.GenerationsCore;
import generations.gg.generations.core.generationscore.client.screen.dialgoue.display.DialogueScreen;
import generations.gg.generations.core.generationscore.network.ClientNetworkPacketHandler;
import generations.gg.generations.core.generationscore.network.packets.GenerationsNetworkPacket;
import net.minecraft.client.Minecraft;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.resources.ResourceLocation;
import org.jetbrains.annotations.NotNull;

import java.util.List;


public record S2CChooseDialoguePacket(String text, List<String> options) implements GenerationsNetworkPacket<S2CChooseDialoguePacket> {

    @Override
    public ResourceLocation getId() {
        return ID;
    }

    @Override
    public void encode(@NotNull FriendlyByteBuf buf) {
        buf.writeUtf(text);
        buf.writeCollection(options, FriendlyByteBuf::writeUtf);
    }

    public static final ResourceLocation ID = GenerationsCore.id("choose_dialogue");
    public static S2CChooseDialoguePacket decode(FriendlyByteBuf buf) {
        return new S2CChooseDialoguePacket(buf.readUtf(), buf.readList(FriendlyByteBuf::readUtf));
    }

}

File: common/src/main/kotlin/generations/gg/generations/core/generationscore/network/packets/dialogue/S2CChooseDialoguePacket.kt
package generations.gg.generations.core.generationscore.network.packets.dialogue

import com.cobblemon.mod.common.api.net.ClientNetworkPacketHandler
import com.cobblemon.mod.common.api.net.NetworkPacket
import generations.gg.generations.core.generationscore.GenerationsCore
import generations.gg.generations.core.generationscore.client.screen.dialgoue.display.DialogueScreen
import generations.gg.generations.core.generationscore.client.screen.dialgoue.display.DialogueScreen.ChooseActiveInfo
import generations.gg.generations.core.generationscore.network.packets.GenerationsNetworkPacket
import net.minecraft.client.Minecraft
import net.minecraft.network.FriendlyByteBuf


class S2CChooseDialoguePacket(private val text: String, private val options: List<String>) : GenerationsNetworkPacket<S2CChooseDialoguePacket> {
    override val id = ID

    override fun encode(buf: FriendlyByteBuf) {
        buf.writeUtf(text)
        buf.writeCollection(options) { obj, value -> obj.writeUtf(value) }
    }

    companion object {
        val ID = GenerationsCore.id("choose_dialogue")
        fun decode(buf: FriendlyByteBuf) = S2CChooseDialoguePacket(buf.readUtf(), buf.readList { obj -> obj.readUtf() })
    }

    class Handler : ClientNetworkPacketHandler<S2CChooseDialoguePacket> {
        override fun handle(packet: S2CChooseDialoguePacket, client: Minecraft) {
            val dialogueScreen = client.screen;

            if (dialogueScreen is DialogueScreen) {
                dialogueScreen.activeInfo = ChooseActiveInfo(packet.text, packet.options)
                dialogueScreen.init(
                    client,
                    dialogueScreen.width,
                    dialogueScreen.height
                ) // reinitialize to get the options widget to pop up
            }
        }

    }
}
